cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/replication/GetRevisionTaskThreaded.java

```diff
@@ -0,0 +1,210 @@
+import java.util.logging.Logger;
+
+/**
+ * Handles calling CouchClient.getDocWithOpenRevisions() for a batch of revisions IDs, with
+ * multiple simultaneous HTTP threads, returning the results back in a manner which can be iterated
+ * over (to avoid deserialising the entire result into memory).
+ * 
+ * For each revision ID, gets the revision tree for a given document ID and lists of open revision IDs
+ * and "atts_since" (revision IDs for which we know we have attachments)
+ *
+ * The document id and open revisions are from one row of change feeds. For example, for the
+ * following change feed:
+ *
+ * {
+ *   "last_seq": 35,
+ *   "results": [
+ *     { "changes":
+ *       [ { "rev": "1-bd42b942b8b672f0289cf3cd1f67044c" } ],
+ *       "id": "2013-09-23T20:50:56.251Z",
+ *       "seq": 27
+ *     },
+ *     { "changes":
+ *       [ { "rev": "29-3f4dabfb32290e557ac1d16b2e8f069c" },
+ *         { "rev": "29-01fcbf8a3f1457eff21e18f7766d3b45" },
+ *         { "rev": "26-30722da17ad35cf1860f126dba391d67" }
+ *       ],
+ *       "id": "2013-09-10T17:47:17.770Z",
+ *       "seq": 35
+ *       }
+ *    ]
+ * }
+ *
+ * For document with id "2013-09-10T17:47:17.770Z", it has open revisions:
+ *
+ *       [
+ *         { "rev": "29-3f4dabfb32290e557ac1d16b2e8f069c" },
+ *         { "rev": "29-01fcbf8a3f1457eff21e18f7766d3b45" },
+ *         { "rev": "26-30722da17ad35cf1860f126dba391d67" }
+ *       ]
+ */
+class GetRevisionTaskThreaded implements Iterable<DocumentRevsList> {
+
+    // TODO logging
+    private static final Logger logger = Logger.getLogger(GetRevisionTaskThreaded.class.getCanonicalName());
+
+    // members used to make requests:
+
+    private final ThreadPoolExecutor executorService;
+    private final ExecutorCompletionService<DocumentRevsList> completionService;
+    private final CouchDB sourceDb;
+    private final List<BulkGetRequest> requests;
+    private final boolean pullAttachmentsInline;
+
+    private boolean iteratorValid = true;
+
+    int threads = 4; // TODO - config?
+
+    // members used to handle responses:
+
+    LinkedBlockingQueue<Future<DocumentRevsList>> responses;
+    AtomicInteger requestsOutstanding;
+
+    public GetRevisionTaskThreaded(CouchDB sourceDb,
+                                   List<BulkGetRequest> requests,
+                                   boolean pullAttachmentsInline) {
+        Preconditions.checkNotNull(sourceDb, "sourceDb cannot be null");
+        Preconditions.checkNotNull(requests, "requests cannot be null");
+        for(BulkGetRequest request : requests) {
+            Preconditions.checkNotNull(request.id, "id cannot be null");
+            Preconditions.checkNotNull(request.revs, "revs cannot be null");
+        }
+
+        this.sourceDb = sourceDb;
+        this.requests = requests;
+        this.pullAttachmentsInline = pullAttachmentsInline;
+        this.executorService = new ThreadPoolExecutor(threads, threads, 1, TimeUnit.MINUTES,
```

Comment:
Given that this `ThreadPoolExecutor` has identical core and max thead number configuration I think this is essentially a fixed size thread pool where the keep alive time has no effect (because it won't reduce below core threads anyway). As such I think `Executors.newFixedThreadPool(threads)` may be more concise.